File: DSA/Graphs/src/com/company/detectCycleUndirected.java
package com.company;

import java.util.ArrayList;
import java.util.LinkedList;
import java.util.Queue;

/**
 * only concern we have in undirected graph is that besides it's immediate parent if it is connected with
 * a vertex which is already visited then it is in a cycle
 */
public class detectCycleUndirected {

    public boolean byDFS(int V, ArrayList<ArrayList<Integer>> adj)
    {
        boolean[] visited = new boolean[V];
        for(int i = 0; i < V; i++) {
            if(!visited[i]) {
                if(DFS(i,adj,visited, -1)) return true;
            }
        }
        return false;
    }

    public boolean DFS(int v, ArrayList<ArrayList<Integer>> adj, boolean[] visited, int parent)
    {
        visited[v] = true;
        for(int u: adj.get(v)) {
            if(!visited[u]) {
                if(DFS(u,adj,visited, v)) return true;

            }
            else if(visited[u] && u != parent) return true;
        }
        return false;
    }
}

File: DSA/Graphs/DetectCycleUndirected.py
class Solution:
    
    #Function to detect cycle in an undirected graph.
	def isCycle(self, V, adj):
		#Code here
		visited = [False] * V
		for i in range(V):
		    if not visited[i]:
		        if self.detect(i,-1,adj,visited):
		            return True
		return False
		
	def detect(self, child, parent, adj, visited):
	    visited[child] = True
	    neighbours = adj[child]
        for nodes in neighbours:
            if visited[nodes] and nodes != parent:
                return True
            elif not visited[nodes]:
	            if self.detect(nodes,child,adj,visited):
	                return True
	                
	    return False
